### PR TITLE
[RW-3586] Enable setting custom environment variables within a cluster

### PR DESCRIPTION
**Pull request by als364**
https://precisionmedicineinitiative.atlassian.net/browse/RW-3586 actual ticket, which probably no one has access to

--

I, the developer opening this PR, do solemnly pinky swear that:

- [x] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoModelCopy.scala
Comment (Qi77Qi) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoModelCopy.scala]: 1. this is probably not the place you want to make the change. https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala#L87 might be the place you need

2. Use `Map[String, String]` instead of wrapping in `Option` since an empty `Map` already has represents the idea of no custom environment variable

**Comment by Qi77Qi** [comment]
what is the motivation for custom environment variable for cluster creation? is this going to be exposed in UI?

**Comment by als364** [comment]
I work on All of Us and we want to expose the workspace CDR version as an environment variable in our notebooks by default.

**Comment by Qi77Qi** [comment]
fyi, https://broadworkbench.atlassian.net/wiki/spaces/IA/pages/100401153/Testing+notebook+functionality+with+Fiab has some documentation around how you can test the change in a fiab

**Comment by als364** [comment]
I don't think I have permissions to view that page.

File: http/src/main/resources/jupyter/init-actions.sh
Comment (rtitle) [inline http/src/main/resources/jupyter/init-actions.sh]: Think this needs to be a bucket path. Also it looks like `jupyter-docker-compose.yaml` is expecting a file named `/etc/custom_env_vars.env`, not `custom-env-vars.json`.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala]: You probably also need to update the json serializer `ClusterRequestFormat`, also in this file. This field can be handled the same way as `labels`.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala]: In `jupyter-docker-compose` it expects a file named `custom_env_vars.env` -- this should probably be consistent.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: It's probably possible to build this content as a `String` and upload to GCS directly, instead of creating a tmp file. It's already in memory in the `clusterRequest`.

File: http/src/main/resources/jupyter/init-actions.sh
Comment (als364) [reply http/src/main/resources/jupyter/init-actions.sh]: Will unify filenames.

Using a bucket path would make more sense. I'm having trouble figuring out whether the bucket we upload things to in `initializeBucketOptions` is passed to this script and if so how.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala
Comment (als364) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala]: Want me to make an `EnvironmentVariableMap` type for this too or just leave it as a `Map[String, String]`?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala
Comment (als364) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: a la the templated resources at line ~970?

File: http/src/main/resources/jupyter/init-actions.sh
Comment (rtitle) [reply http/src/main/resources/jupyter/init-actions.sh]: There's a few things to do. You can probably use `jupyterNotebookConfigUri` as a model.

1. Add a `customEnvVarsUri` field to [ClusterResourcesConfig](https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/ClusterResourcesConfig.scala)
2. Add a `customEnvVars` field to [ClusterInitValues](https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala#L271)
3. In `ClusterInitValues.apply`, populate it like [this](https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala#L341)
4. In [LeonardoService.initializeBucketObjects](https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala#L922), build the content and upload to GCS by adding a call like [this](https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala#L971).
5. In [init-actions.sh](https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/resources/jupyter/init-actions.sh), add a line like [this](https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/resources/jupyter/init-actions.sh#L131)
6. In [init-actions.sh](https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/resources/jupyter/init-actions.sh), add a `gsutil cp` like [this](https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/resources/jupyter/init-actions.sh#L239-L242)
7. Set the `env_file` in jupyter-docker-compose as you have currently

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala]: Nah, leave it as `Map[String, String]`. I don't think the type alias adds any value.

Btw, you can probably just write `Map.empty` - it should infer the types.

Comment (als364) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala]: Scala magic!

File: http/src/main/resources/jupyter/init-actions.sh
Comment (als364) [reply http/src/main/resources/jupyter/init-actions.sh]: For #5, what actually populates the values on the RHS?

Comment (rtitle) [reply http/src/main/resources/jupyter/init-actions.sh]: Leo will replace templated variables in the format `$(..)`. The templated variable corresponds to the variable names in `ClusterInitValues` (steps 2-3).

**Comment by rtitle** [comment]
Btw, you probably need to merge in https://github.com/DataBiosphere/leonardo/pull/1105, or rebase your branch once that PR is merged to fix a compilation error.

**Comment by als364** [comment]
retest this please

**Comment by als364** [comment]
bah

**Comment by als364** [comment]
jenkins retest

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookCustomizationSpec.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookCustomizationSpec.scala]: Could this use the same cluster as above? It's costly to create a lot of clusters in tests; better to reuse them if possible.
```
      withNewCluster(billingProject, request = defaultClusterRequest.copy(customClusterEnvironmentVariables = Map("KEY" -> "value"))) { cluster =>
        withWebDriver { implicit driver =>
          withNewNotebook(cluster, Python3) { notebookPage =>
            notebookPage.executeCell("import os")
            val envVar = notebookPage.executeCell("os.getenv('KEY')")
            envVar shouldBe Some("'value'")
          }
          withNewNotebook(cluster, RKernel) { notebookPage =>
            val envVar = notebookPage.executeCell("Sys.getenv('KEY')")
            envVar shouldBe Some("'value'")
          }
        }
      }
```

**Review by rtitle — CHANGES_REQUESTED** [review]
Looks good! Just 1 minor comment in the test. Other than that 💯

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookCustomizationSpec.scala
Comment (als364) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookCustomizationSpec.scala]: Yes, can do

**Review by rtitle — APPROVED** [review]
:shipit: once tests pass

**Review by Qi77Qi — APPROVED** [review]

**Comment by als364** [comment]
jenkins retest

**Comment by als364** [comment]
@rtitle @Qi77Qi The swatomation test failure happened twice in a row which makes me think it's not just some flakey thing. I have no idea what [this failure](https://fc-jenkins.dsp-techops.broadinstitute.org/job/leonardo-fiab-test-runner/17659/testReport/junit/org.broadinstitute.dsde.workbench.leonardo.notebooks/NotebookCustomizationSpec/NotebookCustomizationSpec_should_populate_user_specified_environment_variables/) means. Have you ever seen something like this before?

**Comment by rtitle** [comment]
I have an idea, can I add a commit to this PR?

**Comment by rtitle** [comment]
I did, let's see how this run does :)

**Comment by rtitle** [comment]
It finally passed! I'll leave it to you @als364 to merge when ready. Remember to Squash and merge.

**Pull request merged**
